File: day01/part2.exs
defmodule Part2 do
  defp parse_line(line) do
    line
    |> String.trim()
    |> String.split("   ")
    |> Enum.map(&String.to_integer/1)
    |> List.to_tuple()
  end

  defp calculate_similarity(left, right) do
    right_count = Enum.frequencies(right)

    left
    |> Enum.map(fn num -> num * Map.get(right_count, num, 0) end)
    |> Enum.sum()
  end

  def solve(filename) do
    filename
    |> File.stream!()
    |> Stream.map(&parse_line/1)
    |> Enum.to_list()
    |> then(fn parsed_lines ->
      left = Enum.map(parsed_lines, &elem(&1, 0))
      right = Enum.map(parsed_lines, &elem(&1, 1))
      calculate_similarity(left, right)
    end)
  end
end

"input.txt"
|> Part2.solve()
|> IO.puts()

File: day1/part2.lua
local function parse_line(line)
  local l, r = line:match("(%d+)%s+(%d+)")
  return tonumber(l), tonumber(r)
end

local function calculate_similarity(left, right)
  local right_count = {}

  for _, num in ipairs(right) do
    right_count[num] = (right_count[num] or 0) + 1
  end

  local total = 0
  for _, num in ipairs(left) do
    total = total + num * (right_count[num] or 0)
  end

  return total
end

local function solve(filename)
  local left = {}
  local right = {}

  for line in io.lines(filename) do
    local l, r = parse_line(line)
    table.insert(left, l)
    table.insert(right, r)
  end

  return calculate_similarity(left, right)
end

print(solve("input.txt"))
